### PR TITLE
Use prepared statements + add concurrency for Cassandra

### PR DESCRIPTION
**Pull request by mieslep**
Inserts now use prepared statements. Batch size is now tunable, as is the number of concurrent requests that can be sent to Cassandra.

### CODE REVIEW

File: langchain/package.json [hunk 1]
```diff
@@ -935,7 +935,7 @@
     "apify-client": "^2.7.1",
```
[the remark this answers is not in the document]
Comment (mieslep) [reply langchain/package.json:935]: The 4.7 driver introduced support for the native vector type. https://github.com/datastax/nodejs-driver/blob/master/CHANGELOG.md

[the remark this answers is not in the document]
Comment (mieslep) [reply langchain/package.json:935]: The 4.7 driver introduced support for the native vector type. https://github.com/datastax/nodejs-driver/blob/master/CHANGELOG.md

File: langchain/src/vectorstores/cassandra.ts [hunk 2]
```diff
@@ -21,7 +21,54 @@ export interface CassandraLibArgs extends DseClientOptions {
...
+  maxConcurrency?: number;
+  batchSize?: number;
+}
+
+/**
+ * Class for managing concurrent tasks, similar to PromisePool (which is
+ * not part of the project's dependencies, and would be preferable to
+ * having this code).
+ */
+class ConcurrencyManager {
```
Comment (jacoblee93) [inline langchain/src/vectorstores/cassandra.ts:34]: We can use the `AsyncCaller` class which I believe would be similar?

https://github.com/langchain-ai/langchainjs/blob/main/langchain/src/embeddings/openai.ts#L246

File: langchain/src/vectorstores/cassandra.ts [hunk 3]
```diff
@@ -63,7 +114,19 @@ export class CassandraStore extends VectorStore {
     this.table = args.table;
     this.primaryKey = args.primaryKey;
     this.metadataColumns = args.metadataColumns;
-    this.indices = args.indices;
+    this.indices = args.indices || [];
+    this.maxConcurrency = Math.floor(args.maxConcurrency || 25);
```
Comment (jacoblee93) [inline langchain/src/vectorstores/cassandra.ts:118]: If we use `AsyncCaller` as suggested above, we can just have the input extend `AsyncCallerParams`: https://github.com/langchain-ai/langchainjs/blob/main/langchain/src/embeddings/base.ts#L7

**Review by jacoblee93 — CHANGES_REQUESTED** [review]
Thank you for this! We have a built-in concurrency utility you can use for consistency.

File: langchain/src/vectorstores/cassandra.ts (hunk 3 above)
Comment (mieslep) [reply langchain/src/vectorstores/cassandra.ts]: done.

Comment (mieslep) [reply langchain/src/vectorstores/cassandra.ts]: done.

**Comment by mieslep** [comment]
@jacoblee93 I've switched over to AsyncCaller, thanks for the pointers to that code.

**Comment by jacoblee93** [comment]
LGTM if it's good with you!

**Review by jacoblee93 — APPROVED** [review]

**Pull request merged**